JitArm64: Release temporary regs with RAII

Sintendo opened the pull request:
We currently call `Unlock` manually everywhere, which is annoying, especially when the register is only sometimes allocated. By adding a small class we can relieve ourselves of this burden.

Just a quick diff to gauge how other devs feel about this.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:
```diff
@@ -155,9 +152,8 @@ void JitArm64::reg_imm(u32 d, u32 a, u32 value, u32 (*do_op)(u32, u32),
-    ARM64Reg WA = gpr.GetReg();
+    auto WA = gpr.GetScopedReg();
     (this->*op)(gpr.R(d), gpr.R(a), value, WA);
-    gpr.Unlock(WA);
```
ComputeRC0 allocates a register, so it's beneficial to unlock this register before calling ComputeRC0.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:
```diff
@@ -544,9 +539,8 @@ void JitArm64::addx(UGeckoInstruction inst)
-    ARM64Reg WA = gpr.GetReg();
+    auto WA = gpr.GetScopedReg();
     ADDI2R(gpr.R(d), gpr.R(in_reg), imm_value, WA);
-    gpr.Unlock(WA);
```
Same here.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:
```diff
@@ -829,11 +821,10 @@ void JitArm64::rlwnmx(UGeckoInstruction inst)
   NEG(WA, gpr.R(b));
   RORV(gpr.R(a), gpr.R(s), WA);
   ANDI2R(gpr.R(a), gpr.R(a), mask, WA);
-  gpr.Unlock(WA);
```
Same here.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:
```diff
@@ -936,9 +926,8 @@ void JitArm64::addic(UGeckoInstruction inst)
-    ARM64Reg WA = gpr.GetReg();
+    auto WA = gpr.GetScopedReg();
     CARRY_IF_NEEDED(ADDI2R, ADDSI2R, gpr.R(d), gpr.R(a), simm, WA);
-    gpr.Unlock(WA);
```
Same here.

JosJuice reviewed (commented):
Overall I think this is a good idea, but you should be careful so you don't accidentally unlock later than before and make the codegen worse. I've left comments on the cases I found. (It only matters when we're register starved, but still.)

Sintendo commented:
Indeed, `ComputeRC0 ` and also `ComputeCarry` may allocate a reg. I've added scopes to solve this. Not sure how I feel about that, it's a bit awkward and maybe non so intuitive to handle this through scoping. I've added an `Unlock` method so you can still explicitly unlock a temporary register, maybe that would be more appropriate to use here. Thoughts?

JosJuice commented:
I'd say I have a slight preference for using scopes the way you're currently doing, but I do also think we should expose an Unlock method in case that's more convenient for any particular use case.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.h:
```diff
@@ -180,6 +180,57 @@ class Arm64RegCache
+    {
+      m_reg = Arm64Gen::ARM64Reg::INVALID_REG;
+      m_gpr = nullptr;
+    }
```
Private functions and variables should be after the public ones, and there should be a `private:` line before them.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:
```diff
@@ -1274,23 +1267,29 @@ void JitArm64::subfex(UGeckoInstruction inst)
     {
-      LoadCarry();
-      CARRY_IF_NEEDED(SBC, SBCS, gpr.R(d), RB, gpr.R(a));
+      auto RB = [&]() -> Arm64GPRCache::ScopedARM64Reg {
```
Using a lambda for this makes the code harder to read than the old approach in my opinion. Using the old approach does mean you have to declare `RB` as `Arm64GPRCache::ScopedARM64Reg` instead of `auto`, but you have to write that out in the lambda return type here anyway.

JosJuice commented:
Okay, looks good.

Since this is marked WIP, does that mean you're planning on continuing and applying this to more files?

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_LoadStore.cpp:
```diff
@@ -939,7 +940,7 @@ void JitArm64::dcbx(UGeckoInstruction inst)
-  gpr.Unlock(WA, WB, physical_addr);
+  gpr.Unlock(WA, WB);
   if (make_loop)
     gpr.Unlock(loop_counter);
```
It would be nice to have a function in the register cache that lets us use scoped registers for fixed registers like these. But I suppose this PR is getting a bit large already, so you don't have to put that in this PR.

Sintendo replied:
Good idea, should be easy. I'll make a PR for that once this lands.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_LoadStorePaired.cpp:
```diff
@@ -173,115 +173,116 @@ void JitArm64::psq_stXX(UGeckoInstruction inst)
-  ARM64Reg VS = fpr.R(inst.RS, have_single ? RegType::Single : RegType::Register);
-
-  if (js.assumeNoPairedQuantize)
   {
```
You don't need to add an additional scope for the sake of unlocking VS earlier. Unlocking it last will just move it past some other unlock operations, not past anything that emits instructions.

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_SystemRegisters.cpp:
```diff
@@ -553,114 +535,112 @@ void JitArm64::crXXX(UGeckoInstruction inst)
-  WB = ARM64Reg::INVALID_REG;
   gpr.BindCRToRegister(field, true);
-  XB = gpr.CR(field);
+  ARM64Reg field_reg = gpr.CR(field);
```
Customary naming:

```suggestion
  ARM64Reg CR = gpr.CR(field);
```

JosJuice reviewed (commented):
Needs a rebase.

Sintendo commented:
Appreciate the reviews, apologies for dragging my feet on this!

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_RegCache.h:
```diff
@@ -180,6 +180,59 @@ class Arm64RegCache
+  public:
+    inline ScopedARM64Reg() = default;
+    ScopedARM64Reg(const ScopedARM64Reg&) = delete;
+    inline ScopedARM64Reg(Arm64RegCache& cache) : m_reg(cache.GetReg()), m_gpr(&cache) {}
```
`explicit`

JosJuice commented on Source/Core/Core/PowerPC/JitArm64/JitArm64_Integer.cpp:
```diff
@@ -1399,16 +1400,11 @@ void JitArm64::subfic(UGeckoInstruction inst)
     // d = imm - a
     ARM64Reg RD = gpr.R(d);
-    ARM64Reg WA = ARM64Reg::WZR;
-    if (!is_zero)
```
What happened to the `is_zero` optimization here?
